File: lab1/uart.c
#include "gpio.h"

#define AUX_ENABLE  ((volatile unsigned int*)(MMIO_BASE + 0x00215004))
#define AUX_MU_IO   ((volatile unsigned int*)(MMIO_BASE + 0x00215040))
#define AUX_MU_IER  ((volatile unsigned int*)(MMIO_BASE + 0x00215044))
#define AUX_MU_IIR  ((volatile unsigned int*)(MMIO_BASE + 0x00215048))
#define AUX_MU_LCR  ((volatile unsigned int*)(MMIO_BASE + 0x0021504C))
#define AUX_MU_MCR  ((volatile unsigned int*)(MMIO_BASE + 0x00215050))
#define AUX_MU_LSR  ((volatile unsigned int*)(MMIO_BASE + 0x00215054))
#define AUX_MU_CNTL ((volatile unsigned int*)(MMIO_BASE + 0x00215060))
#define AUX_MU_BAUD ((volatile unsigned int*)(MMIO_BASE + 0x00215068))

void uart_init()
{
    *AUX_ENABLE |= 1;
    *AUX_MU_CNTL = 0;
    *AUX_MU_IER = 0;
    *AUX_MU_LCR = 3;
    *AUX_MU_MCR = 0;
    *AUX_MU_BAUD = 270;
    *AUX_MU_IIR = 6;
    *GPFSEL1 &= ~((7 << 12) | (7 << 15));
    *GPFSEL1 |= (2 << 12) | (2 << 15);
    
    *GPPUD = 0;
    for(int i = 0; i < 150; i++) asm volatile("nop");
    *GPPUDCLK0 = (1 << 14) | (1 << 15);
    for(int i = 0; i < 150; i++) asm volatile("nop");
    *GPPUDCLK0 = 0;
    
    *AUX_MU_CNTL = 3;
}

char uart_getc()
{
    char c;
    while ( !(*AUX_MU_LSR&0x01) );
    c = (char)(*AUX_MU_IO);
    if ( c == '\r') { c = '\n'; }
    return c;
}

void uart_send(unsigned int c)
{
    while( !(*AUX_MU_LSR&0x20) );
    *AUX_MU_IO = c;
}

void uart_puts(char *s)
{
    while(*s)
    {
        if(*s == '\n')
        {
            uart_send('\r');
        }
        uart_send(*s++);
    }
}

File: impl-rs/src/uart.rs
use crate::cpu;
use core::ptr::{read_volatile, write_volatile};

const MMIO_BASE: u32 = 0x3F00_0000;

const AUX_ENABLE: *mut u32 = (MMIO_BASE + 0x0021_5004) as *mut u32;
const AUX_MU_IO: *mut u32 = (MMIO_BASE + 0x0021_5040) as *mut u32;
const AUX_MU_IER: *mut u32 = (MMIO_BASE + 0x0021_5044) as *mut u32;
const AUX_MU_IIR: *mut u32 = (MMIO_BASE + 0x0021_5048) as *mut u32;
const AUX_MU_LCR: *mut u32 = (MMIO_BASE + 0x0021_504C) as *mut u32;
const AUX_MU_MCR: *mut u32 = (MMIO_BASE + 0x0021_5050) as *mut u32;
const AUX_MU_LSR: *mut u32 = (MMIO_BASE + 0x0021_5054) as *mut u32;
const AUX_MU_CNTL: *mut u32 = (MMIO_BASE + 0x0021_5060) as *mut u32;
const AUX_MU_BAUD: *mut u32 = (MMIO_BASE + 0x0021_5068) as *mut u32;

const GPFSEL1: *mut u32 = (MMIO_BASE + 0x00200004) as *mut u32;
const GPPUD: *mut u32 = (MMIO_BASE + 0x00200094) as *mut u32;
const GPPUDCLK0: *mut u32 = (MMIO_BASE + 0x00200098) as *mut u32;

pub fn init_uart() {
    unsafe {
        write_volatile(AUX_ENABLE, read_volatile(AUX_ENABLE) | 1);
        write_volatile(AUX_MU_CNTL, 0);
        write_volatile(AUX_MU_LCR, 3);
        write_volatile(AUX_MU_MCR, 0);
        write_volatile(AUX_MU_IER, 0);
        write_volatile(AUX_MU_IIR, 0xc6);
        write_volatile(AUX_MU_BAUD, 270);
        /* map UART1 to GPIO pins */
        let mut r = read_volatile(GPFSEL1);
        r &= !((7 << 12) | (7 << 15));
        r |= (2 << 12) | (2 << 15);
        write_volatile(GPFSEL1, r);
        write_volatile(GPPUD, 0);
        cpu::spin_for_cycles(150);
        write_volatile(GPPUDCLK0, (1 << 14) | (1 << 15));
        cpu::spin_for_cycles(150);
        write_volatile(GPPUDCLK0, 0);
        write_volatile(AUX_MU_CNTL, 3);
    }
}

pub fn getc() -> char {
    unsafe {
        while 0 == (read_volatile(AUX_MU_LSR) & 0x01) {
            cpu::nop();
        }
        let c = read_volatile(AUX_MU_IO) as u8 as char;
        if c == '\r' {
            return '\n';
        }
        return c;
    }
}

pub fn send(c: char) {
    unsafe {
        while 0 == (read_volatile(AUX_MU_LSR) & 0x20) {
            cpu::nop();
        }
        write_volatile(AUX_MU_IO, c as u32);
    }
}
